### PR TITLE
PMP: Add function that trims triangle mesh at a close polyline

### PR DESCRIPTION
**Pull request by afabri**
## Summary of Changes

We add a function that trims a triangle mesh at a given polyline that is close to the surface. 
The current implementation projects the points and then creates a clipping mesh by extruding 
the projected polyline by local normals. 

## Todo
* [ ] The extrusion distance is still a constant
* [ ] Is the function name well chosen since we keep both parts 
* [ ] Could this be implemented with the shortest paths via projections points (intrinsic triangulation ?).

## Release Management

* Affected package(s): Polygon_mesh_processing
* Issue(s) solved (if any): -
* Feature/Small Feature (if any):  tbd
* Link to compiled documentation:  [link](https://cgal.github.io/6862/v0/Polygon_mesh_processing/group__PMP__corefinement__grp.html#ga5024bb7302c9a7159c7f1dc1b5e3b0b8)
* License and copyright ownership: GeometryFactory

### CODE REVIEW

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/trim.h [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+  typedef CGAL::AABB_face_graph_triangle_primitive<TriangleMesh>          AABB_face_graph_primitive;
+  typedef CGAL::AABB_traits<Geom_traits, AABB_face_graph_primitive>       AABB_face_graph_traits;
+  typedef CGAL::AABB_tree<AABB_face_graph_traits>                         Tree;
+  typedef Face_location<TriangleMesh, double>                        Face_location;
+
+  Tree tree;
+  build_AABB_tree(tm, tree);
+
+  int n = std::distance(pr.begin(), pr.end());
+  const double extrusion_distance = 0.1;
```
Comment (sloriot) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/trim.h:71]: this looks arbitrary to me

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/trim.h [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+  // We could have used a modulo for closing
+  {
+    std::array<int,3> a = {2*(n-1)-2, 2*(n-1)-1, 0};
+    faces.push_back(a);
+  }
+  {
+    std::array<int,3> a = {2*(n-1)-1, 0, 1};
+    faces.push_back(a);
+  }
```
Comment (sloriot) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/trim.h:102]:
```suggestion
    faces.push_back({2*(n-1)-2, 2*(n-1)-1, 0});
    faces.push_back({2*(n-1)-1, 0, 1});
```

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/trim.h [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+  // The extruded polyline may have problems
+  bool orientable = orient_polygon_soup(points,faces);
+  if(! orientable){
+    std::cerr << "Provide better curve" << std::endl;
+    return false;
+  }
```
Comment (sloriot) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/trim.h:109]: shouldn't you check for self-intersections too?

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/trim.h [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+  // CGAL::IO::write_polygon_mesh("splitter.off", splitter);
+
+  split(tm, splitter);
+
+  std::vector<TriangleMesh> cctm;
+
+  split_connected_components(tm, cctm);
+
+  if(cctm.size() != 2){
```
Comment (sloriot) [inline Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/trim.h:122]: why? it should be one component more than the input

File: Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/trim.h (hunk 1 above)
Comment (sloriot) [reply Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/trim.h:71]: What you can do is keep the normal direction per vertex and increase the extrusion while the trimer mesh does not intersect entirely the mesh. You can check that for each "quad" corresponding to a segment of the polyline

**Comment by MaelRL** [comment]
After offline discussion, it was decided to postpone the integration of this new function as the current implementation is too tailored for specific inputs (single CC, closed mesh, nice polyline, etc.) compared to the genericity of the name.

Also the projection of the polyline onto the mesh should likely be geodesics-based instead.

**Comment by sloriot** [comment]
Could it be a demo plugin?

**Pull request closed without merging**